packages/zoe/src/contractSupport/types.js

```diff
@@ -49,3 +49,73 @@
  * @param {string} [rightHasExitedMsg]
  * @returns {string}
  */
+
+/**
+ * @typedef {Object} OfferToReturns
+ *
+ * The return value of offerTo is a promise for the userSeat for the
+ * offer to the other contract, and a promise (`deposited`) which
+ * resolves when the payout for the offer has been deposited to the `toSeat`
+ * @property {Promise<UserSeat>} userSeatPromise
+ * @property {Promise<AmountKeywordRecord>} deposited
+ */
+
+/**
+ * @typedef {Record<Keyword,Keyword>} KeywordKeywordRecord
+ *
+ * A mapping of keywords to keywords.
+ */
+
+/**
+ * @callback OfferTo
+ *
+ * Make an offer to another contract instance (labeled contractB below),
+ * withdrawing the payments for the offer from a seat in the current
+ * contract instance (contractA) and depositing the payouts in another
+ * seat in the current contract instance (contractA).
+ *
+ * @param {ContractFacet} zcf
+ *   Zoe Contract Facet for contractA
+ *
+ * @param {ERef<Invitation>} invitation
+ *   Invitation to contractB
+ *
+ * @param {KeywordKeywordRecord=} keywordMapping
+ *   Mapping of keywords used in contractA to keywords to be used in
+ *   contractB. Note that the pathway to deposit the payout back to
+ *   contractA reverses this mapping.
+ *
+ * @param {Proposal} proposal
+ *   The proposal for the offer to be made to contractB
+ *
+ * @param {ZCFSeat} fromSeat
+ *   The seat in contractA to take the offer payments from.
+ *
+ * @param {ZCFSeat} toSeat
+ *   The seat in contractA to deposit the payout of the offer to.
+ *
+ * @returns {OfferToReturns}
+ */
+
+/**
+ * @callback Reverse
+ *
+ * Given a mapping of keywords to keywords, invert the keys and
+ * values. This is used to map the offers made to another contract
+ * back to the keywords used in the first contract.
+ * @param {KeywordKeywordRecord=} keywordRecord
+ * @returns {KeywordKeywordRecord }
+ */
+
+/**
+ * @callback MapKeywords
+ *
+ * Remap the keywords of an amountKeywordRecord or a
+ * PaymentPKeywordRecord according to a mapping. This is used to remap
+ * from keywords used in contractA to keywords used in contractB and
+ * vice versa in `offerTo`
+ *
+ * @param {AmountKeywordRecord | PaymentPKeywordRecord | undefined }
+ * keywordRecord
+ * @param {KeywordKeywordRecord} keywordMapping
```

Comment:
shouldn't this have an `@returns` clause?

Reply:
Ah good catch, thanks!

Reply:
Hmm, I just tried this and if I say that the return value is {AmountKeywordRecord | PaymentPKeywordRecord | undefined }, it is not smart enough to know that the same kind will be returned. I don't know how to specify this. But with this return value, I get errors in `offerTo` saying that I can't pass something potentially undefined to the other methods that use the result of `mapKeywords`